br/pkg/backup/client.go

```diff
@@ -871,570 +812,380 @@ func (bc *Client) BackupRanges(
-	// Keys are saved in encoded format in TiKV, so the key must be encoded
-	// in order to find the correct region.
-	var leader *metapb.Peer
-	key = codec.EncodeBytesExt([]byte{}, key, isRawKv)
-	state := utils.InitialRetryState(60, 100*time.Millisecond, 2*time.Second)
-	failpoint.Inject("retry-state-on-find-target-peer", func(v failpoint.Value) {
-		logutil.CL(ctx).Info("reset state for FindTargetPeer")
-		state = utils.InitialRetryState(v.(int), 100*time.Millisecond, 100*time.Millisecond)
-	})
-	err := utils.WithRetry(ctx, func() error {
-		region, err := bc.mgr.GetPDClient().GetRegion(ctx, key)
-		failpoint.Inject("return-region-on-find-target-peer", func(v failpoint.Value) {
-			switch v.(string) {
-			case "nil":
-				{
-					region = nil
-				}
-			case "hasLeader":
-				{
-					region = &pd.Region{
-						Leader: &metapb.Peer{
-							Id: 42,
-						},
-					}
-				}
-			case "hasPeer":
-				{
-					region = &pd.Region{
-						Meta: &metapb.Region{
-							Peers: []*metapb.Peer{
-								{
-									Id:      43,
-									StoreId: 42,
-								},
-							},
-						},
-					}
-				}
+func (bc *Client) OnBackupResponse(
+	ctx context.Context,
+	r *ResponseAndStore,
+	errContext *utils.ErrorContext,
+	globalProgressTree *rtree.ProgressRangeTree,
+) error {
+	if r == nil || r.GetResponse() == nil {
+		return nil
+	}
 
-			case "noLeader":
-				{
-					region = &pd.Region{
-						Leader: nil,
-					}
-				}
-			case "noPeer":
-				{
-					{
-						region = &pd.Region{
-							Meta: &metapb.Region{
-								Peers: nil,
-							},
-						}
-					}
-				}
-			}
-		})
-		if err != nil || region == nil {
-			logutil.CL(ctx).Error("find region failed", zap.Error(err), zap.Reflect("region", region))
-			return errors.Annotate(berrors.ErrPDLeaderNotFound, "cannot find region from pd client")
+	resp := r.GetResponse()
+	storeID := r.GetStoreID()
+	if resp.GetError() == nil {
+		pr, err := globalProgressTree.FindContained(resp.StartKey, resp.EndKey)
+		if err != nil {
+			logutil.CL(ctx).Error("failed to update the backup response",
+				zap.Reflect("error", err))
+			return err
 		}
-		if len(targetStoreIds) == 0 {
-			if region.Leader != nil {
-				logutil.CL(ctx).Info("find leader",
-					zap.Reflect("Leader", region.Leader), logutil.Key("key", key))
-				leader = region.Leader
-				return nil
-			}
-		} else {
-			candidates := make([]*metapb.Peer, 0, len(region.Meta.Peers))
-			for _, peer := range region.Meta.Peers {
-				if _, ok := targetStoreIds[peer.StoreId]; ok {
-					candidates = append(candidates, peer)
-				}
+		if bc.checkpointRunner != nil {
+			if err := checkpoint.AppendForBackup(
+				ctx,
+				bc.checkpointRunner,
+				pr.GroupKey,
+				resp.StartKey,
+				resp.EndKey,
+				resp.Files,
+			); err != nil {
+				// flush checkpoint failed,
+				// this error doesn't not influnce main procedure. so ignore it.
+				logutil.CL(ctx).Warn("failed to flush checkpoint, ignore it", zap.Error(err))
```

Comment:
maybe need to adjust checkpoint to make it not stop itself after send error back.